<a href="https://colab.research.google.com/github/eckoecho/CodingDojo/blob/week-10/Week%2010/PCA_Exercise_(Core).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [20]:
import pandas as pd
import numpy as np
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.pipeline import make_pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Load Data

In [21]:
#Loading images from sklearn library
mnist = fetch_openml("mnist_784")

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


In [22]:
#Accessing the data
mnist.data.head()

pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  pixel9  \
0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

   pixel10  ...  pixel775  pixel776  pixel777  pixel778  pixel779  pixel780  \
0      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
1      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
2      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
3      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
4      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   

   pixel781  pixel782  pixel783  pixel784  
0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0  

[5 rows x 784 columns]

In [23]:
#Accesing and establishing features
X = mnist.data
X.head()

pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  pixel9  \
0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

   pixel10  ...  pixel775  pixel776  pixel777  pixel778  pixel779  pixel780  \
0      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
1      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
2      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
3      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
4      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   

   pixel781  pixel782  pixel783  pixel784  
0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0  

[5 rows x 784 columns]

In [24]:
#Accessing and assigning target
y = mnist.target
y.head()

0    5
1    0
2    4
3    1
4    9
Name: class, dtype: category
Categories (10, object): ['0', '1', '2', '3', ..., '6', '7', '8', '9']

# Cleaning data

In [25]:
#Checking class weight
y.value_counts(normalize=True)

1    0.112529
7    0.104186
3    0.102014
2    0.099857
9    0.099400
0    0.098614
6    0.098229
8    0.097500
4    0.097486
5    0.090186
Name: class, dtype: float64

In [26]:
#Checking for duplicates
X.duplicated().sum()

0

In [27]:
#Checking for null values
X.isna().sum().sum()

0

# Train Test Split

# Model 1: KNN w/ PCA

In [28]:
#Performing train test split
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=42)
#Checking values
X_train.head()

pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  pixel9  \
58774     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
7414      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
63662     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
54551     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
50995     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

       pixel10  ...  pixel775  pixel776  pixel777  pixel778  pixel779  \
58774      0.0  ...       0.0       0.0       0.0       0.0       0.0   
7414       0.0  ...       0.0       0.0       0.0       0.0       0.0   
63662      0.0  ...       0.0       0.0       0.0       0.0       0.0   
54551      0.0  ...       0.0       0.0       0.0       0.0       0.0   
50995      0.0  ...       0.0       0.0       0.0       0.0       0.0   

       pixel780  pixel781  pixel782  pixel783  pixel784  
58774       0.0       0.0       0.0       0.0       0.0  
7414        0.0       0.0       0.0       0.0       0.0  
63662       0.0       0.0       0.0       0.0       0.0  
54551       0.0       0.0       0.0       0.0       0.0  
50995       0.0       0.0       0.0       0.0       0.0  

[5 rows x 784 columns]

##Creating Scaler

In [29]:
#instantiate scaler
scaler = StandardScaler()

## Using Scaled Data w/ PCA

## Pipeline for Scaler

In [30]:
#Instantiate knn
# make pipe with scaler and knn model
scaler_pca_pipe = make_pipeline(scaler, PCA(.95))

In [31]:
# Putting scaler and PCA in a pipeline
transformer = make_pipeline(StandardScaler(), scaler_pca_pipe)

In [32]:
#putting scaler combined with knn
knn_pca_pipe= make_pipeline(transformer, KNeighborsClassifier())
#Fitting training data with KNN, Scaling, PCA
knn_pca_pipe.fit(X_train, y_train)
knn_pca_preds = knn_pca_pipe.predict(X_test)

#Get accuracy

In [33]:
#When predicting categorical values you can use the accuracy_score
print(f"KNN w/ PCA Accuracy Score: {accuracy_score(y_test, knn_pca_preds)}")

KNN w/ PCA Accuracy Score: 0.9477714285714286


# Model 2: KNN w/ Scaled Data

In [38]:
#Instantiate KNN Model
knn = KNeighborsClassifier()
#Fitting with scaler and the KNN model
knn_pipe = make_pipeline(scaler, knn)
knn_pipe.fit(X_train, y_train)
knn_test_preds = knn_pipe.predict(X_test)
#When predicting categorical values you can use the accuracy_score
print(f"KNN w/ Scaling Accuracy Score {accuracy_score(y_test, knn_test_preds)}")

KNN w/ Scaling Accuracy Score 0.9442285714285714


# Evaluate and compare the models

In [39]:
%%time
knn_pca_preds = knn_pca_pipe.predict(X_test)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


CPU times: user 44.4 s, sys: 235 ms, total: 44.6 s
Wall time: 26.5 s


In [40]:
%%time
knn_preds = knn_pipe.predict(X_test)

CPU times: user 1min 36s, sys: 169 ms, total: 1min 36s
Wall time: 58.2 s


a. Which model performed the **best on the test** set?

>The accuracy with for the KNN model using PCA had the best prediction at **94.77%** as opposed to not using PCA with an accuracy score of **94.42%**.

b. Which model was the **fastest at making predictions**?

>KNN using PCA is the fastest as well.  It executes in **26 seconds** as opposed to not using PCA with a execution time of **58 seconds**.